lib/ndebug/ndebug.c

```diff
@@ -0,0 +1,190 @@
+    0x02,           /* endpoint count */
+    0xff,           /* interface class - User Deficned */
+    0x00,           /* interface subclass - Patched by client */
+    0x00,           /* interface protocol - Patched by client */
+    0x00,           /* string index */
+
+    /* endpoint 1 IN */
+    0x07,           /* length */
+    ENDPOINT,       /* type */
+    0x80,           /* address - Patched by Client */
+    0x02,           /* type: bulk */
+    W(64),          /* max packet size: 64 */
+    00,             /* interval - unused for bulk */
+
+    /* endpoint 1 OUT */
+    0x07,           /* length */
+    ENDPOINT,       /* type */
+    0x00,           /* address - Patched by client */
+    0x02,           /* type: bulk */
+    W(64),          /* max packet size: 64 */
+    00,             /* interval - unused for bulk */
+};
+
+static void init_channel(uint8_t subclass, uint8_t protocol, uint8_t ep_addr)
+{
+    uint8_t desc[sizeof(bulk_pair_descriptor_template)];
+
+    // Make a copy of the template descriptor and fill in the missing fields.
+    memcpy(desc, bulk_pair_descriptor_template,
+           sizeof(bulk_pair_descriptor_template));
+
+    desc[IFACE_SUBCLASS_IDX] = subclass;
+    desc[IFACE_PROTOCOL_IDX] = protocol;
+    desc[IFACE_IN_ADDR_IDX] = ep_addr | 0x80;
+    desc[IFACE_OUT_ADDR_IDX] = ep_addr;
+
+    // Append the interfaces.
+    usb_append_interface_lowspeed(desc, sizeof(desc));
+    usb_append_interface_highspeed(desc, sizeof(desc));
+}
+
+static void setup_usb_endpoint(uint8_t ep_num)
+{
+    usbc_setup_endpoint(ep_num, USB_IN, 0x40, USB_BULK);
+    usbc_setup_endpoint(ep_num, USB_OUT, 0x40, USB_BULK);
+}
+
+static status_t ndebug_register_cb(
+    void *cookie,
+    usb_callback_op_t op,
+    const union usb_callback_args *args
+)
+{
+    if (op == USB_CB_ONLINE) {
+        for (channel_t ch = 0; ch < NDEBUG_CHANNEL_COUNT; ++ch) {
+            setup_usb_endpoint(CH_TO_ADDR(ch));
+        }
+        event_signal(&usb_online_event, false);
+    } else if (op == USB_CB_RESET || op == USB_CB_DISCONNECT ||
+               op == USB_CB_OFFLINE) {
+        event_unsignal(&usb_online_event);
+    }
+    return NO_ERROR;
+}
+
+void ndebug_init(void)
+{
+    for (channel_t ch = 0; ch < NDEBUG_CHANNEL_COUNT; ++ch) {
+        event_init(&rx_event[ch], 0, EVENT_FLAG_AUTOUNSIGNAL);
+        event_init(&tx_event[ch], 0, EVENT_FLAG_AUTOUNSIGNAL);
+    }
+
+    init_channel(NDEBUG_SUBCLASS, NDEBUG_PROTOCOL_LK_SYSTEM, SYS_EP_ADDR);
+    init_channel(NDEBUG_SUBCLASS, NDEBUG_PROTOCOL_SERIAL_PIPE, USR_EP_ADDR);
+
+    event_init(&usb_online_event, 0, 0);
+
+    usb_register_callback(&ndebug_register_cb, NULL);
+}
+
+static status_t usb_xmit_cplt_cb(ep_t endpoint, usbc_transfer_t *t)
+{
+    uint32_t channel = (uint32_t)t->extra;
+    CHECK_CHANNEL(channel);
+
+    event_signal(&tx_event[channel], false);
+    return 0;
+}
+
+static status_t usb_recv_cplt_cb(ep_t endpoint, usbc_transfer_t *t)
```

Comment:
should probably check the result of the transfer and make sure it actually received something. IIRC the len field is adjusted with the amount of data actually received.